### DISCUSSION TITLE
Allow to run on airplane mode, please!

### DISCUSSION THREAD

**Discussion by gitanO0** [NightscoutFoundation/xDrip#3270] [General]
Google has relaxed how it implements airplane mode and it seems like xdrip+ has become even more strict. 
Google allows airplane mode to be enabled on Android devices while keeping Bluetooth turned on. 
Xdrip+ has had this limitation where it won't run if airplane mode is turned on.

**Comment by Navid200** [comment]
Please explain what you mean.
xDrip keeps working if airplane mode is active or not.  

There are phones that automatically disable Bluetooth when you enable airplane mode.  In that case, you will need to reenable Bluetooth after enabling airplane mode since xDrip will need Bluetooth to act as a collector.  But, it's not xDrip that disables Bluetooth in this case.  It is the phone.
xDrip itself does not care if airplane mode is on or not.

**Comment by gitanO0** [comment]
It's not the enabling/disabling of Bluetooth that's the issue.  The issue is that if airplane mode is enabled, xdrip will stop collecting data and says so in the event log.  Once airplane mode is disabled again, xdrip starts working again.  If what you say is true that "xDrip keeps working if airplane mode is active or not", then something is very wrong with my setup because xdrip is telling me in the event log, "Not setting Bluetooth to state:  true due to airplane mode being enabled".

**Reply by Navid200** [reply]
Please post a screenshot of the log you are talking about.
Also, please tell us if you are using an xDrip that is not from this repository.  If you are using xDrip from this repository, please tell us which version.

**Comment by gitanO0** [comment]
To clarify further... my Bluetooth remains enabled and on even when I turn on Airplane Mode... that's the "relaxing" that Google has done lately.  They've made it so that even if you enable airplane mode, wifi, and Bluetooth will stay enabled.  Xdrip on the other hand doesn't respect that and just stops working simply because airplane mode is turned on.

**Comment by gitanO0** [comment]
![Screenshot_20240103-201542](https://github.com/NightscoutFoundation/xDrip/assets/6887761/94f523ca-6ee8-4e3b-bd5c-0ea687ba63d6)

**Comment by gitanO0** [comment]
Also, of note.  Over the years, I've been able to use airplane mode and have Xdrip work... for a while, but eventually, I'll get the issue stated above.  Lately, however, as soon as I enable airplane mode, the collection stops cold.  I use airplane mode all the time, for all kinds of reasons.  To have xdrip monitor that setting and not whether Bluetooth is actually on and running seems pretty questionable.  I'd ask that the app simply monitor Bluetooth and not give a single care about the airplane mode setting.

**Reply by Navid200** [reply]
After you put your phone in airplane mode, go to your phone's settings and make sure Bluetooth is enabled.
xDrip will work fine and will never disable Bluetooth because the phone is in airplane mode.
I have flown with xDrip and I have never had xDrip stop working.

**Reply by gitanO0** [reply]
I've had it this way for a long time.  I don't know what I could be doing wrong.   I've turned airplane mode on many, many times and bluetooth stays enabled.  That has never been the issue.

**Reply by gitanO0** [reply]
The issue has been explicitly, that when airplane mode is on and with Bluetooth on, functioning and enabled, xdrip stops working and gives the event logs above.  I guess the question is what is different about my set up than yours?

**Reply by Navid200** [reply]
I just checked the code.
My understanding is the following.

By default, xDrip automatically enables Bluetooth if it needs Bluetooth.  So, if you just disable Bluetooth on your phone, xDrip will reenable it.  
However, if airplane mode is on and Bluetooth is disabled, when xDrip goes to enable it, it will not enable Bluetooth and will give that message.

There is nowhere in xDrip that Bluetooth is disabled because airplane mode is on.

xDrip disables Bluetooh as a part of watchdog.
So, I can see a consequence of that and I am happy to investigate and make changes if necessary.

This is what I suspect you have experienced:
1- You enabled airplane mode.
2- Bluetooth was active.
3- There were multiple disconnects, bad enough that the Bluetooth watchdog was triggered.
4- Bluetooth watchdog disables Bluetooth and then reenables it.
5- This is my suspicion and I need to investigate to confirm.  If I am right, when Bluetooth watchdog goes to reenable Bluetooth, it does not because airplane mode is on.

If this is true, I agree that we have a problem that needs to be fixed.  
Please give me time to consult with the lead developer.

**Reply by gitanO0** [reply]
Ah, that all makes sense.  I think what happens with me is that for example, I'm at home, my phone is in airplane mode but with wifi and BT enabled (I have shitty cell reception at my house).  I then walk down to my basement while my phone is upstairs and the connection is lost.  Xdrip then goes into its "BT watchdog" thing and bam, the issue happens.  
Does that jive with what you are saying?

**Reply by Navid200** [reply]
Yes.  But, this is only a suspicion.
I would be very surprised if this has never come up so far.  I suspect I will be told that will never happen.
That's why I said I needed to consult and investigate.

**Reply by gitanO0** [reply]
I do believe this has been brought up in the past.  A few folks, including me, have asked about this, albeit quite some time ago for me.  The response has been a blanket "The airplane mode limitation will not be modified".  I searched the discussions and came up with zero results for "airplane"... I did see one or two issues on the matter.  Also, it might have been that I communicated with Jamorham directly via gitter.

**Reply by gitanO0** [reply]
Something has changed with this over the past day or two... not sure if it's an Android update, which just happened today (Jan Android 14 update) or an update to xdrip code but now, if I put my phone into airplane mode, it won't complete even one collection... I can have the phone on my person the entire time, so it doesn't look like this is specifically related to loss of connection, BT Watchdog.  It looks like it's purely associated with airplane mode.  In the past, I could go for seemingly days before this issue would crop up.  Now, it's as soon as I turn on airplane mode.  I've verified that BT is still on and functioning so I have no idea what is triggering this but as of now,  I can no longer use airplane mode whatsoever and expect xdrip to function.

**Comment by Navid200** [comment]
I went over the watchdog code.
It doesn't care about airplane mode.  So, I was wrong.  This cannot be caused by airplane mode.

Would you please tell me what your phone brand and model is?

This is a separate note:
Would you please explain why you use airplane mode when you are not flying?

**Reply by gitanO0** [reply]
Verizon Pixel 7 Pro Android 14, January '24 update.

I use airplane mode all the time (probably more on airplane mode than not).  If I'm at work or home, it's on.  Why?  Because my cell connectivity at my house and work stinks, so leaving my cell data on all day needlessly drains my battery.  Why don't I turn off "mobile data" instead of turning on airplane mode?  Two reasons:  1)  Google has made it so that when airplane mode is on, you can keep BT and WIFI enabled so, toggling airplane mode doesn't affect BT or WIFI connectivity.  2)  Enabling airplane mode to turn off mobile data (cell) is FAR EASIER than turning off mobile data.  To enable/disable airplane mode, I simply tap the "airplane mode" button in quick settings.  To toggle mobile data, I have to go to settings, internet, cell, and then toggle "mobile data".  Since I do this multiple times a day, the former option is far easier.  

I think the notion that "airplane mode" is only relevant when on an airplane is very unrealistic.  It's a simple, convenient way to disable cell connectivity thus saving battery while also keeping BT and Wifi active.  Here's a [good overview](https://www.sammobile.com/news/android-14-smarter-flight-mode-samsung-galaxy-phones/) of the newer Android functionality.  

I guess I could just take this approach as well:  I travel a lot, and I'm on planes a lot.  Every time I activate airplane mode, xdrip+ stops collecting.  On long flights, this is very problematic and tempts me to turn off airplane mode.  My BT and WIFI are still on and active during the flights, which is ok because the planes have wifi so why xdrip can't work in that kind of situation is very unfortunate. 

Thanks for looking into this.  Not sure why I'm such an edge case regarding this topic.  I'd think a lot of folks would be hip to this usage pattern.

**Reply by Navid200** [reply]
I have been able to recreate the problem.
I need to figure out why this happens.  So that I don't forget, please click on "Create issue from discussion" in the right pane.

Please change the title to something like:
Bluetooth may be disabled by airplane mode

**Reply by gitanO0** [reply]
Ok, thanks @Navid200 !  Appreciate you looking into this.  Will create the issue ASAP.
